domain/cloud/state/state.go

```diff
@@ -73,6 +75,258 @@ func (st *State) ListClouds(ctx context.Context, filter *Filter) ([]cloud.Cloud,
+				return fmt.Errorf("missing cloud %q %w%w", cloudName, errors.NotValid, errors.Hide(err))
+			} else if err != nil {
+				return fmt.Errorf("updating cloud default keys %q: %w", cloudName, err)
+			}
+		}
+
+		return nil
+	})
+}
+
+// CloudAllRegionDefaults returns all the default settings for a cloud and it's
+// regions. Note this will not include the defaults set on the cloud itself but
+// just that of it's regions. Empty map values are returned when no region
+// defaults are found.
+func (st *State) CloudAllRegionDefaults(
+	ctx context.Context,
+	cloudName string,
+) (map[string]map[string]string, error) {
+	defaults := map[string]map[string]string{}
+
+	db, err := st.DB()
+	if err != nil {
+		return defaults, fmt.Errorf("getting database instance for cloud region defaults: %w", err)
+	}
+
+	stmt := `
+SELECT cloud_region.name,
+       cloud_region_defaults.key,
+       cloud_region_defaults.value
+FROM cloud_region_defaults
+INNER JOIN cloud_region
+ON cloud_region.uuid = cloud_region_defaults.region_uuid
+INNER JOIN cloud
+ON cloud_region.cloud_uuid = cloud.uuid
+WHERE cloud.name = ?
+`
+
+	return defaults, db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, stmt, cloudName)
+		if err != nil {
+			return fmt.Errorf("fetching cloud %q region defaults: %w", cloudName, err)
+		}
+
+		var regionName, key, value string
+		for rows.Next() {
+			if err := rows.Scan(&regionName, &key, &value); err != nil {
+				return fmt.Errorf(
+					"compiling cloud %q region %q defaults: %w",
+					cloudName,
+					regionName,
+					stderrors.Join(err, rows.Close()),
+				)
+			}
+			store, has := defaults[regionName]
+			if !has {
+				store = map[string]string{}
+				defaults[regionName] = store
+			}
+			store[key] = value
+		}
+		return nil
+	})
+}
+
+// UpdateCloudRegionDefaults is responsible for updating default config values
+// for a cloud region. This function will allow the addition and updating of
+// attributes. Attributes can also be removed by keys if they exist for the
+// current cloud. If the cloud or region is not found an error that satisfied
+// NotValid is returned.
+func (st *State) UpdateCloudRegionDefaults(
+	ctx context.Context,
+	cloudName string,
+	regionName string,
+	updateAttrs map[string]string,
+	removeAttrs []string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteBinds, deleteVals := database.SliceToPlaceholder(removeAttrs)
+	deleteStmt := fmt.Sprintf(`
+DELETE FROM cloud_region_defaults
+WHERE key IN (%s)
+AND region_uuid = (SELECT cloud_region.uuid
+                   FROM cloud_region
+                   INNER JOIN cloud
+                   ON cloud_region.cloud_uuid = cloud.uuid
+                   WHERE cloud.name = ?
+                   AND cloud_region.name = ?)`, deleteBinds)
+
+	upsertBinds, upsertVals := database.MapToMultiPlaceholder(updateAttrs)
```

Comment:
We can swing around on this later.

To me it's very opaque for a simple operation - I can see why we do the temp table thing in order to incorporate the join for cloud and region UUIDs, but my hope would be that we're handing around objects at a higher level such that we can parameterise directly with the IDs we need.

For controller config (no other entity IDs needing to be filled) I just looped inside the transaction. For something that isn't a hot path, the clarity is worth it.

Reply:
Sounds good.